src/SlackFunction.ts

```diff
@@ -46,83 +113,412 @@ export class SlackFunction {
   private callbackId: string;
 
   /**
-   * @description fn to to process corresponding
+   * @description handler to to process corresponding
    * function_executed event
    */
-  private fn: Middleware<SlackEventMiddlewareArgs>;
+  private handler: Middleware<SlackEventMiddlewareArgs>;
+
+  private interactivityHandlers: FunctionInteractivityMiddleware[];
+
+  private logger: Logger;
+
+  public constructor(callbackId: string, handler: Middleware<SlackEventMiddlewareArgs>) {
+    validate(callbackId, handler);
 
-  public constructor(callbackId: string, fn: Middleware<SlackEventMiddlewareArgs>) {
-    // TODO: Add validation step
     this.callbackId = callbackId;
-    this.fn = fn;
+    this.handler = handler;
+    this.interactivityHandlers = [];
+
+    // set an initial default logging
+    const logger = new ConsoleLogger();
+    logger.setName(`SlackFunction: [${this.callbackId}]`);
+    logger.setLevel(LogLevel.DEBUG);
+    this.logger = logger;
+  }
+
+  /**
+   * Attach a block_actions interactivity handler to your SlackFunction
+   *
+   * ```
+   * Example:
+   * const actionHandler = async () => {};
+   * const actionHandler1 = async () => {};
+   * myFunc.action("id", actionHandler).action("id1", actionHandler1);
+   * ```
+   *
+   * @param actionIdOrConstraints Provide an action_id string
+   * corresponding to the value supplied in your blocks or a
+   * constraint object of type ActionConstraints<SlackAction>
+   *
+   * ```
+   * Example:
+   * myFunc.action({ type: "action_submission" });
+   * myFunc.action({ action_id: "id" }, actionHandler);
+   * ```
+   * @param handler Provide a handler function
+   * @returns SlackFunction instance
+   */
+  public action<
+    Action extends SlackAction = SlackAction,
+    Constraints extends ActionConstraints<Action> = ActionConstraints<Action>,
+  >(
+    actionIdOrConstraints: string | RegExp | Constraints,
+    handler: Middleware<SlackActionMiddlewareArgs>,
+  ): this {
+    // normalize constraints
+    const constraints: ActionConstraints = (
+      typeof actionIdOrConstraints === 'string' ||
+      util.types.isRegExp(actionIdOrConstraints)
+    ) ?
+      { action_id: actionIdOrConstraints } :
+      actionIdOrConstraints;
+
+    // declare our valid constraints keys
+    const validConstraintsKeys: ActionConstraintsKeys = ['action_id', 'block_id', 'callback_id', 'type'];
```

Comment:
These constraints keys types are here to help us make sure we're never declaring validConstraintsKeys that are not corresponding to the underlying ActionConstraints or ViewsConstraints types.